code/modules/projectiles/guns/_shared_ammo_objects.dm

```diff
@@ -0,0 +1,201 @@
+// Used for objects created by ammo datums that tend to be more general and can be used in multiple ammo datum files.
+
+
+/////////////////////////////
+//          FIRE           //
+/////////////////////////////
+// Base type , meant to be overridden
+/obj/fire_base
+	name = "fire"
+	desc = "Ouch!"
+	anchored = TRUE
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	icon = 'icons/effects/fire.dmi'
+	icon_state = "red_2"
+	layer = BELOW_OBJ_LAYER
+	/// tracks for how many process ticks the fire will exist.Can be reduced by other sources
+	var/burn_ticks = 10
+
+/obj/fire_base/Initialize(mapload)
+	. = ..()
+	START_PROCESSING(SSobj, src)
+
+	var/static/list/connections = list(
+		COMSIG_ATOM_ENTERED = PROC_REF(on_cross),
+	)
+	AddElement(/datum/element/connect_loc, connections)
+
+/obj/fire_base/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+///Effects applied to anything that crosses a burning turf
+/obj/fire_base/proc/on_cross(datum/source, atom/movable/crosser, oldloc, oldlocs)
+	SIGNAL_HANDLER
+	return
+
+/// Effects applied from smokes
+/obj/fire_base/effect_smoke(obj/effect/particle_effect/smoke/affecting_smoke)
+	return
+
+/// Called on all objects on a turf on process(). Doesn't include the turf itself
+/obj/fire_base/proc/affect_mob(mob/living/carbon/affected)
+	return
+
+/// Called on the turf we are currently burning
+/obj/fire_base/proc/affect_turf(turf/affected)
+	return
+
+/// Called on all objs on the turf we are on
+/obj/fire_base/proc/affect_obj(obj/affected)
+	return
+
+/obj/fire_base/process()
+	var/turf/burn_turf = loc
+	if(!istype(burn_turf)) //Is it a valid turf?
+		qdel(src)
+		return
+
+	burn_ticks--
+	if(burn_ticks <= 0)
+		qdel(src)
+		return PROCESS_KILL
+
+	affect_turf(burn_turf)
+	for(var/atom/thing AS in burn_turf)
+		if(isliving(thing))
+			affect_mob(thing)
+			continue
+		if(isobj(thing))
+			affect_obj(thing)
+			continue
+
+/////////////////////////////
+//      FLAMER FIRE        //
+/////////////////////////////
+
+/obj/fire_base/fire
+	icon_state = "red_2"
```

Comment:
This is way too set for pyrogen fire, like
A) fire_base is a bad type name use something like fire
B) effect_smoke copy pasted twice instead of using parent
C) On cross code is straight up copy pasted twice basically
D) set_fire is not on the parent, this was half the reason so you dont manually copy paste and use the same structure yes i know it has flamer fire act I told you to split that section into a different proc
E) stuff like particles/lighting defined on subtypes instead of being on parent and just toggling light_on

Basically go through and make sure the code is actually shared properly